### PR TITLE
Enhancements in Documentation and Code Readability

### PR DESCRIPTION
**Pull request by cbielow**
## Description

This PR primarily focuses on fixing Doxygen warnings in our C++ codebase and thus improving the documentation for about 80 classes. It also includes some code refactoring and optimization, particularly in the OPXLHelper and SimpleTSGXLMS classes. 

Additionally, a new test is added to check for Doxygen warnings in the future and automatically fail CI to prevent warnings from building up.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

___

## **Type**
Enhancement, Bug fix

___

## **Description**
This PR includes a variety of changes, primarily focused on improving documentation and code readability. Some of the most significant changes include:
- Reordering and adding some includes in `DefaultParamHandlerDocumenter.cpp` to improve the documentation generation process.
- Removing an unnecessary critical section and refactoring a function in `OPXLHelper.cpp` to improve readability and performance.
- Updating the copyright year and fixing references in the documentation of `PeptideIndexer.cpp` and `ProteinQuantifier.cpp`.
- Adding a new file `Doxygen_Warning_Checker.cpp` that adds a test to check for doxygen warnings, ensuring that the documentation is generated without issues.
- Fixing a condition to process a spectrum in `OPXLSpectrumProcessingAlgorithms.cpp`, which was previously ignored due to a misplaced condition.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th>Relevant files&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>DefaultParamHandlerDocumenter.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        doc/doxygen/parameters/DefaultParamHandlerDocumenter.cpp<br><br>

**The changes in this file involve reordering and adding some <br>includes, and modifying the way some objects are <br>instantiated for documentation. The changes aim to improve <br>the documentation generation process.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-8b2d4d82a4fc7c64b362f58dc68ca60aa3a40d3a8c8cf009009d452f10738f85"> +17/-14</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PeptideIndexer.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/topp/PeptideIndexer.cpp<br><br>

**The changes in this file involve updating the copyright year <br>and fixing a reference in the documentation.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-4e0926b5fcf84e9b873ba7ab7c45b5ed8a5cc815ea1e218ef6a1f7324c2d77c4"> +2/-4</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ProteinQuantifier.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/topp/ProteinQuantifier.cpp<br><br>

**The changes in this file involve fixing a reference in the <br>documentation.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-4599045b736f2b99f91b28709cd376fe762963336f0ad5b0387015e2af00460b"> +1/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>RNPxlXICFilter.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/topp/RNPxlXICFilter.cpp<br><br>

**The changes in this file involve updating a reference in the <br>documentation.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-69cabac94e26c32dad45218c7f130cbcb1b2c402117d2854453a19c37299517f"> +1/-1</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>OPXLSpectrumProcessingAlgorithms.h&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/openms/include/OpenMS/ANALYSIS/XLMS/OPXLSpectrumProcessingAlgorithms.h<br><br>

**The changes in this file involve updating the documentation <br>for a function to make it more clear and accurate.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-49f4cf723ef6106ee58f14c8ef4e6a1b91aa4df6622a270964db236e16862f24"> +14/-16</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>OPXLHelper.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/openms/source/ANALYSIS/XLMS/OPXLHelper.cpp<br><br>

**The changes in this file involve removing an unnecessary <br>critical section, and refactoring a function to improve <br>readability and performance.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-10b3f96b2eaaea682b8c2a6813787e919cfe0d5e7bd6a6a9a7dd8e9b8543ef96"> +20/-46</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>Doxygen_Warning_Checker.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        doc/doxygen/test/Doxygen_Warning_Checker.cpp<br><br>

**This is a new file that adds a test to check for doxygen <br>warnings. This will help ensure that the documentation is <br>generated without issues.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-6bff4d80e6a460de16fb49601a8e34e846c24dfe061bb7258b270871c30b36bc"> +41/-0</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>OPXLSpectrumProcessingAlgorithms.cpp&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/openms/source/ANALYSIS/XLMS/OPXLSpectrumProcessingAlgorithms.cpp<br><br>

**The changes in this file involve fixing a condition to <br>process a spectrum, which was previously ignored due to a <br>misplaced condition.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-f2a9ca55b5f6372eb315aa5e6c89d7e8e6605ae449b067f54ee9f25c54201766"> +2/-2</a></td>

</tr>                    
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>sources.cmake&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        src/openms/include/OpenMS/ANALYSIS/TARGETED/sources.cmake<br><br>

**The changes in this file involve reordering some includes. <br>This does not affect the functionality of the code.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7295/files#diff-d60e2f13b74c2224be7e6c9aa971f6fbef1fbe8f326327ac3fb500f0e44ee03e"> +2/-2</a></td>

</tr>                    
</table></details></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

**Comment by cbielow** [comment]
ok, this PR now works as intended, but if new doxygen warnings are introduced in the future, the build will fail like this:
```
Generating docs for file ChromatogramPeak.h...
Generating docs for file ChromatogramSettings.h...
Generating docs for file ChromatogramTools.h...
Generating docs for file ChromeleonFile.h...
Generating docs for file ClassTest.h...
Exiting...
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 255.
```

with nothing on CDash.
Also the actual doxygen error is not shown.

This is a little inconvenient.
Any good ideas to solve this?
Shall we have test which checks the doxygen-error.log instead and reports its content?
This would allow to fix future doxygen warnings without actually running doxygen locally.
Also, the user would see all warnings. Right now, you only get the first one, since doxygen is instructed to fail after it encounters the first warning ATM.

**Comment by timosachsenberg** [comment]
I think you need to add the extra test.

**Comment by timosachsenberg** [comment]
@CodiumAI-Agent /review

**Comment by timosachsenberg** [comment]
just a test ;)

**Comment by cbielow** [comment]
ok, the new test seems to work:
https://cdash.openms.de/test/131499345

**Comment by cbielow** [comment]
will go ahead and fix the remaining warnings. After that, ready for merge.

**Review by timosachsenberg — APPROVED** [review]

**Comment by timosachsenberg** [comment]
@cbielow the AI updated the PR description. Agree that this is useful?

**Comment by cbielow** [comment]
Agreed.

However, the AI code analysis definitely has some false-positives.
e.g.
```
The function addXLinkIonPeaks_ has been refactored to use const reference for its peptide parameter. However, the function's documentation has not been updated to reflect this change. Please update the function's documentation accordingly. [important]
```
, yet all I did was remove a whitespace…

**Comment by cbielow** [comment]
can someone guess why the ubuntu package builds now fail with
```
CMake Error at CMakeLists.txt:328 (is_valid_package):

  is_valid_package Macro invoked with incorrect arguments for macro named:
  is_valid_package
```
when CMake is re-invoked after the docs are build?
The failing line of code is here: https://github.com/OpenMS/OpenMS/blob/develop/CMakeLists.txt#L328

**Comment by jpfeuffer** [comment]
I guess if pkg type is not set, the action will put a "" there. but only "none" is allowed.
=> set the pkg type always and do not rely on the default:
https://github.com/OpenMS/OpenMS/actions/runs/7626626098/workflow?pr=7295#L85

File: .github/workflows/openms_ci_matrix_full.yml [hunk 1]
```diff
@@ -77,12 +77,12 @@ jobs:
...
           echo "xvfb=xvfb-run -a" >> $GITHUB_OUTPUT
           echo "static_boost=OFF" >> $GITHUB_OUTPUT
+          # always run doxygen on Ubuntu (because its fast), to detect Doxygen errors
+          echo "enable_docs=ON" >> $GITHUB_OUTPUT  
           if [ "$DO_PACKAGE" = true ]; then
-            echo "enable_docs=ON" >> $GITHUB_OUTPUT
             echo "pkg_type=deb" >> $GITHUB_OUTPUT
           else
             echo "enable_docs=OFF" >> $GITHUB_OUTPUT
-            echo "pkg_type=none" >> $GITHUB_OUTPUT
```
Comment (jpfeuffer) [inline .github/workflows/openms_ci_matrix_full.yml:85]: => I basically mean this

**Comment by cbielow** [comment]
I think I found a solution (passing the PACKAGE_TYPE as ENV variable), but this is only required because
https://github.com/OpenMS/OpenMS/blob/develop/tools/ci/cibuild.cmake#L139
seems to delete the old Cache entirely.
Is this really required?

**Comment by cbielow** [comment]
closing, since merging is not possible, for some reason. An exact clone of this PR has no problems though (#7313).
Will continue over there.

**Comment by jpfeuffer** [comment]
I think the package step uses cipackage.cmake, no?

**Comment by jpfeuffer** [comment]
Yes, everything is passed as ENV variable. That's the idea.
Would be great to have CMake configs or how they are called but creating a huge "-D" command line is even uglier IMO.

**Comment by cbielow** [comment]
> I think the package step uses cipackage.cmake, no?

yes, but building uses cibuild.cmake (https://github.com/OpenMS/OpenMS/blob/develop/.github/workflows/openms_ci_matrix_full.yml#L284) and also resets the Cache.

**Comment by jpfeuffer** [comment]
But building is only run in the very beginning isn't it? So there should not be any cache yet anyways

**Comment by jpfeuffer** [comment]
But yes. We probably don't need to delete the cache. Just a safety measure I guess.

**Comment by cbielow** [comment]
yes, but for some reason, CMake reconfigures itself after running `ninja doc` now, and then the `PACKAGE_TYPE` was empty.
I don't fully understand why, because the first time CMake runs, it somehow has the `PACKAGE_TYPE` set.

**Comment by jpfeuffer** [comment]
During build:
- cibuild is used, which does NOT set PACKAGE_TYPE through an ENV variable => internal OpenMS CMake default is used => "none"

During package:
- cipackage is used. CMake is reconfigured (as always/usual)
- PACKAGE_TYPE is set as an ENV var by the GitHub action
- You removed the pkg_type output variable which is used for PACKAGE_TYPE. This output variable will be empty
- PACKAGE_TYPE will be ""
- CMake reconfigure will fail since it needs to be "none" or a valid package type.

Solution:
Always set pkg_type output variable OR handle empty vars in the cipackage.cmake
Or we need to handle it more gracefully in our internal CMake, since "none" is weird.

**Pull request closed without merging**